File: usage-daemon/src/providers/cohere.js
// Cohere usage provider plugin — descriptor interface.
//
// Cohere's public API has no usage/quota endpoint (confirmed: /check-api-key
// only validates key identity). The dashboard.cohere.com "Usage" tab calls an
// internal RPC instead:
//
//   POST https://production.api.os.cohere.com/rpc/BlobheartAPI/GetAPIUsage
//     Authorization: Bearer <dashboard session JWT>
//     body: { req: { orgID, before, after, includeTotal: true } }
//     -> { usages: [{ productID, productName, productUnit, quantity, time, total }] }
//
// This is a per-model/per-day token ledger, NOT a used/cap quota — Cohere's
// documented trial limit ("1,000 API calls a month", docs.cohere.com/docs/
// rate-limits) is a call COUNT, but this endpoint only reports token
// quantities, so there's no way to derive "X of 1000 calls" from it. The
// window below is informational only: total input+output tokens summed over
// a trailing 30-day span, no cap, no pct, no resets_at.
//
// Auth: the Bearer token is a dashboard SESSION JWT pulled from browser
// storage (sessionStorage/localStorage), not a cookie — cookiejar.js only
// reads cookies.sqlite, so cookie_from_firefox CANNOT recover this one. It
// must be pasted manually (auth kind 'token') and re-pasted when it expires
// (observed lifespan ~5 days: exp - iat). org_id is not secret — it's the
// orgID visible in the same dashboard session/network calls.
//
// parse() is a PURE function of the envelope JSON text.

export const TOKENS_COLOR = '#D55E00'; // Okabe-Ito vermillion

export const ID = 'cohere';
export const LABEL = 'Cohere';

const USAGE_URL = 'https://production.api.os.cohere.com/rpc/BlobheartAPI/GetAPIUsage';
const USER_AGENT = 'usage-daemon/0.1';
const WINDOW_DAYS = 30;

export class AuthExpiredError extends Error {
  constructor(msg = 'Cohere dashboard session token missing or expired') {
    super(msg);
    this.code = 'auth_expired';
  }
}

export class RateLimitedError extends Error {
  constructor(retryAfter = null) {
    super('rate_limited');
    this.code = 'rate_limited';
    this.retryAfter = retryAfter;
  }
}

// Pure function of the envelope JSON text — no fs/network.
// envelope: { usages: [{ productUnit, quantity, ... }] }
export function parse(raw) {
  let env;
  try {
    env = typeof raw === 'string' ? JSON.parse(raw) : raw;
  } catch {
    throw new AuthExpiredError('unparseable cohere usage envelope');
  }
  if (!env || typeof env !== 'object' || !Array.isArray(env.usages)) {
    throw new AuthExpiredError('unparseable cohere usage envelope');
  }

  let inputTokens = 0;
  let outputTokens = 0;
  for (const u of env.usages) {
    const qty = typeof u.quantity === 'number' ? u.quantity : 0;
    if (u.productUnit === 'Input tokens') inputTokens += qty;
    else if (u.productUnit === 'Output tokens') outputTokens += qty;
  }
  const totalTokens = inputTokens + outputTokens;

  const windows = [
    {
      id: 'tokens',
      label: `Tokens (${WINDOW_DAYS}d)`,
      letter: 'Tk',
      pct: null, // no documented cap for token volume — informational only
      used: totalTokens,
      cap: null,
      unit: 'tokens',
      resets_at: null, // rolling window, not a billing-cycle reset
      color: TOKENS_COLOR,
      will_deplete: false,
    },
  ];

  return {
    tier: null,
    windows,
    segments: [],
    // carried out of parse so meta() can surface the raw figures too.
    _cohere: { input_tokens: inputTokens, output_tokens: outputTokens, total_tokens: totalTokens },
  };
}

function createProvider() {
  let token = null;
  let orgId = null;
  let lastStats = null; // surfaced via meta()

  return {
    id: ID,
    label: LABEL,
    auth: { kind: 'token' },

    config() {
      return {
        id: ID,
        label: LABEL,
        usageUrl: 'https://dashboard.cohere.com/usage',
        // path/relogin: the pasted value is a short-lived dashboard session
        // JWT (~5 days), not the long-lived API key — re-paste from
        // dashboard.cohere.com network traffic (Authorization header) when
        // it expires. No auto-refresh path exists (not a browser cookie).
        auth: { kind: 'token', relogin: 're-paste dashboard session JWT (~5d lifespan)' },
        // Support/inference service usage ledger, informational only — no
        // quota cap is derivable from this endpoint.
        category: 'support',
        windows: [{ id: 'tokens', label: `Tokens (${WINDOW_DAYS}d)`, color: TOKENS_COLOR }],
        tiers: [],
      };
    },

    configure(cfg = {}) {
      // !== undefined so configure({api_token:''}) can explicitly clear it.
      if (cfg.api_token !== undefined) token = cfg.api_token ? String(cfg.api_token).trim() : null;
      if (cfg.org_id !== undefined) orgId = cfg.org_id ? String(cfg.org_id).trim() : null;
    },

    async setAuth(payload) {
      token = String(payload ?? '').trim() || null;
    },

    async fetch() {
      if (!token) throw new AuthExpiredError('no Cohere dashboard session token configured');
      if (!orgId) throw new Error('cohere: no org_id configured (see config.example.toml)');

      const now = new Date();
      const after = new Date(now.getTime() - WINDOW_DAYS * 24 * 60 * 60 * 1000);

      const res = await fetch(USAGE_URL, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
          'User-Agent': USER_AGENT,
          Accept: '*/*',
          'Request-Source': 'playground',
          Origin: 'https://dashboard.cohere.com',
          Referer: 'https://dashboard.cohere.com/',
        },
        body: JSON.stringify({
          req: { orgID: orgId, before: now.toISOString(), after: after.toISOString(), includeTotal: true },
        }),
      });

      if (res.status === 401 || res.status === 403) throw new AuthExpiredError();
      if (res.status === 429) {
        throw new RateLimitedError(Number(res.headers.get('retry-after')) || null);
      }
      if (!res.ok) throw new Error(`production.api.os.cohere.com HTTP ${res.status}`);

      const body = await res.json();
      // Cache figures for meta() (parse() computes them too, but the runner
      // calls meta() independently of parse()).
      lastStats = parse(JSON.stringify(body))._cohere;
      return JSON.stringify(body);
    },

    intervalSeconds() {
      return 900; // billing ledger, no rush — poll less often than a live quota
    },

    meta() {
      return lastStats
        ? {
            input_tokens: lastStats.input_tokens,
            output_tokens: lastStats.output_tokens,
            total_tokens: lastStats.total_tokens,
          }
        : {};
    },

    parse,
  };
}

export { createProvider };

File: usage-daemon-v3/usage_daemon/providers/cohere.py
"""Cohere usage provider plugin (port of src/providers/cohere.js, extended per
the locked data model).

Cohere's public API has no usage/quota endpoint. The dashboard "Usage" tab calls
an internal RPC:

  POST https://production.api.os.cohere.com/rpc/BlobheartAPI/GetAPIUsage
    Authorization: Bearer <dashboard session JWT>
    body: { req: { orgID, before, after, includeTotal: true } }
    -> { usages: [{ productID, productName, productUnit, quantity, time, total }] }

This is a per-model/per-day token ledger. LOCKED extensions over the JS:
  - tokens window over a trailing month bucket (WINDOW_DAYS=30, matches both the
    JS code and the live dashboard request Aug 5 -> Sep 5).
  - NEW `calls` window read from the RESPONSE HEADERS:
      x-ratelimit-limit: 300, x-ratelimit-remaining, x-ratelimit-reset
    used = limit - remaining, cap = limit, pct = used/cap. This is the real
    quota the JS ignored.
  - per-model segments parsed out of productName ("... on <Model> model").
  - meta gains a per-day burn curve + rate-limit state.

Auth is a pasted dashboard session JWT (~5-day lifespan), not a browser cookie —
cookie_from_firefox cannot recover it.
"""

from __future__ import annotations

import json
import re

from ..errors import AuthExpiredError
from ..httputil import create_client


TOKENS_COLOR = "#D55E00"
CALLS_COLOR = "#D55E00"

ID = "cohere"
LABEL = "Cohere"

USAGE_URL = "https://production.api.os.cohere.com/rpc/BlobheartAPI/GetAPIUsage"
USER_AGENT = "usage-daemon"
WINDOW_DAYS = 30

_RE_MODEL = re.compile(r"on (?:default )?(.+?) model", re.I)


def _parse_model(product_name: str) -> str:
    m = _RE_MODEL.search(product_name or "")
    return m.group(1).strip() if m else product_name


def _calls_window(headers):
    def _h(name: str):
        if not headers:
            return None
        v = headers.get(name) or headers.get(name.lower()) or headers.get(name.title())
        return v

    limit = _h("x-ratelimit-limit")
    remaining = _h("x-ratelimit-remaining")
    reset = _h("x-ratelimit-reset")
    try:
        limit = int(limit)
    except (TypeError, ValueError):
        limit = None
    try:
        remaining = int(remaining)
    except (TypeError, ValueError):
        remaining = None
    if limit is None or remaining is None:
        return None
    used = max(0, limit - remaining)
    return {
        "id": "calls",
        "label": "Calls",
        "letter": "Ca",
        "pct": (100 * used) / limit,
        "used": used,
        "cap": limit,
        "unit": "calls",
        "resets_at": reset,  # epoch seconds; runner's to_host_iso renders it
        "color": CALLS_COLOR,
        "will_deplete": False,
    }


def parse(raw, headers: dict | None = None) -> dict:
    """Pure function of the envelope JSON text (+ optional response headers)."""
    try:
        env = json.loads(raw) if isinstance(raw, str) else raw
    except Exception:
        raise AuthExpiredError("unparseable cohere usage envelope")
    if not env or not isinstance(env, dict) or not isinstance(env.get("usages"), list):
        raise AuthExpiredError("unparseable cohere usage envelope")

    input_tokens = 0
    output_tokens = 0
    per_model: dict[str, dict] = {}
    daily: dict[str, dict] = {}
    for u in env["usages"]:
        qty = u.get("quantity") if isinstance(u.get("quantity"), (int, float)) else 0
        unit = u.get("productUnit")
        if unit == "Input tokens":
            input_tokens += qty
        elif unit == "Output tokens":
            output_tokens += qty
        model = _parse_model(u.get("productName"))
        pm = per_model.setdefault(
            model, {"model": model, "input_tokens": 0, "output_tokens": 0, "total_tokens": 0}
        )
        key = "input_tokens" if unit == "Input tokens" else ("output_tokens" if unit == "Output tokens" else None)
        if key:
            pm[key] += qty
            pm["total_tokens"] = pm["input_tokens"] + pm["output_tokens"]
        day = (u.get("time") or "")[:10]
        if day:
            d = daily.setdefault(day, {"date": day, "input": 0, "output": 0, "total": 0})
            if key == "input_tokens":
                d["input"] += qty
            elif key == "output_tokens":
                d["output"] += qty
            d["total"] = d["input"] + d["output"]

    total_tokens = input_tokens + output_tokens
    windows = [
        {
            "id": "tokens",
            "label": f"Tokens ({WINDOW_DAYS}d)",
            "letter": "Tk",
            "pct": None,  # informational — no quota cap for token volume
            "used": total_tokens,
            "cap": None,
            "unit": "tokens",
            "resets_at": None,  # rolling window — the API exposes no reset date
            "color": TOKENS_COLOR,
            "will_deplete": False,
        }
    ]
    calls = _calls_window(headers)
    if calls:
        windows.append(calls)

    return {
        "tier": None,
        "windows": windows,
        "segments": sorted(per_model.values(), key=lambda s: -s["total_tokens"]),
        "_cohere": {
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "total_tokens": total_tokens,
            "daily": sorted(daily.values(), key=lambda d: d["date"]),
            "rate_limit": (
                {"limit": calls["cap"], "remaining": calls["cap"] - calls["used"], "reset_at": calls["resets_at"]}
                if calls
                else None
            ),
        },
    }


def create_provider(client=None):
    st = {"token": None, "orgId": None, "lastHeaders": None, "client": client}

    def config():
        return {
            "id": ID,
            "label": LABEL,
            "usageUrl": "https://dashboard.cohere.com/usage",
            "auth": {
                "kind": "token",
                "relogin": "re-paste dashboard session JWT (~5d lifespan)",
            },
            "category": "support",
            "windows": [
                {"id": "tokens", "label": f"Tokens ({WINDOW_DAYS}d)", "color": TOKENS_COLOR},
                {"id": "calls", "label": "Calls", "color": CALLS_COLOR},
            ],
            "tiers": [],
        }

    def configure(cfg: dict | None = None):
        cfg = cfg or {}
        if "api_token" in cfg:
            st["token"] = cfg["api_token"].strip() if cfg["api_token"] else None
        if "org_id" in cfg:
            st["orgId"] = cfg["org_id"].strip() if cfg["org_id"] else None

    async def set_auth(payload: str) -> None:
        st["token"] = (payload or "").strip() or None

    async def fetch() -> str:
        if not st["token"]:
            raise AuthExpiredError("no Cohere dashboard session token configured")
        if not st["orgId"]:
            raise RuntimeError("cohere: no org_id configured (see config.example.toml)")

        from datetime import datetime, timedelta, timezone

        now = datetime.now(timezone.utc)
        after = now - timedelta(days=WINDOW_DAYS)
        body = json.dumps({
            "req": {
                "orgID": st["orgId"],
                "before": now.isoformat(),
                "after": after.isoformat(),
                "includeTotal": True,
            }
        })
        c = st["client"] or create_client()
        res = await c.post(
            USAGE_URL,
            headers={
                "Authorization": f"Bearer {st['token']}",
                "Content-Type": "application/json",
                "User-Agent": USER_AGENT,
                "Accept": "*/*",
                "Request-Source": "playground",
                "Origin": "https://dashboard.cohere.com",
                "Referer": "https://dashboard.cohere.com/",
            },
            content=body,
        )
        if res.status_code in (401, 403):
            raise AuthExpiredError()
        if res.status_code == 429:
            from ..errors import RateLimitedError

            ra = res.headers.get("retry-after")
            raise RateLimitedError(int(ra) if ra and ra.isdigit() else None)
        if res.status_code >= 400:
            raise RuntimeError(f"production.api.os.cohere.com HTTP {res.status_code}")
        st["lastHeaders"] = dict(res.headers)
        return res.text

    def parse_fn(raw):
        return parse(raw, headers=st["lastHeaders"])

    def interval_seconds():
        return 900  # billing ledger — poll less often than a live quota

    def meta():
        # parse_fn() is where the runner gets the stats; surface a static label
        # hint so clients know the ledger's window without re-fetching.
        return {"window_days": WINDOW_DAYS}

    return {
        "id": ID,
        "label": LABEL,
        "auth": {"kind": "token"},
        "config": config,
        "configure": configure,
        "set_auth": set_auth,
        "fetch": fetch,
        "interval_seconds": interval_seconds,
        "meta": meta,
        "parse": parse_fn,
    }